[JS] August release notes (#1629)
Azure/azure-sdk: releases/2020-08/js.md

The Azure SDK team is pleased to make available the August 2020 client library release.

#### Updates

- Core libraries.
- Identity.
- Azure Event Hubs.
- Azure Form Recognizer.
- Azure Key Vault.

#### Preview

> - Azure Service Bus.

Reviewer: @HarshaNalluru , @ramya-rao-a  - anything to add here?

Reviewer 2: Added service bus changelog

## Installation Instructions

To install the packages, copy and paste the below into a terminal.

```bash
$> npm install @azure/identity
$> npm install @azure/eventhubs-checkpointstore-blob
$> npm install @azure/ai-form-recognizer
$> npm install @azure/keyvault-keys
$> npm install @azure/keyvault-secrets
$> npm install @azure/keyvault-certificates

> $> npm install @azure/service-bus

Reviewer 3: Service bus is preview while the latest is GA 1.1.8 so need `@next`
```suggestion
$> npm install @azure/service-bus@next
```

The Azure SDK team is pleased to make available the August 2020 client library release.

> #### GA

Author: Thanks! I'll ask today.

> #### GA

Author: Cala recommends moving them to GA. I'm doing so.

> #### Updates

Author: Gotcha! I'll ask today.

> #### Updates

Author: @deyaaeldeen mentioned that the relevant change was this:

> Updated the package version to be 5.0.0 instead of 1.0.0 in order to reduce confusion with older versions of the Azure Text Analytics SDKs.

Then you mentioned to me that if that's the only change, we don't need to call it out. So I'll resolve this item.

- Core libraries.
- Identity.
- Azure Event Hubs.
- Azure Form Recognizer.
- Azure Key Vault.

#### Preview

- Azure Service Bus.

## Installation Instructions

To install the packages, copy and paste the below into a terminal.

```bash
$> npm install @azure/identity
$> npm install @azure/eventhubs-checkpointstore-blob
$> npm install @azure/ai-form-recognizer
$> npm install @azure/keyvault-keys
$> npm install @azure/keyvault-secrets
$> npm install @azure/keyvault-certificates
$> npm install @azure/service-bus@next
```

## Feedback

If you have a bug or feature request for one of the libraries, please post an issue at the [azure-sdk-for-js repository](https://github.com/azure/azure-sdk-for-js/issues).

## Changelog

Detailed changelogs are linked from the [Quick Links](#quick-links) below. Here are some of the highlights:

### Identity

> #### Identity's [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

Author: Thank you, I'll ask.

> #### Identity's [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

Author: I confirmed with Cala that this should be `Identity Changelog`. I'll do that change.

> #### Identity's [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

Author: I followed your suggestion. Cala confirmed this approach. Other languages are also using `Identity Changelog` style too. This change was already made! I'll resolve this feedback.

- The terms `RuntimeInfo` and `Description` have been replaced with `RuntimeProperties` and `Properties` to better align with guidelines around the kind of suffixes we use for naming methods and interfaces.

##### New Features

- User agent details can now be added to the outgoing requests by passing the user-agent prefixes to the `ServiceBusClient` and the `ServiceBusManagementClient` through options.
  Example user-agent string if the prefix `SampleApp` is provided to `ServiceBusManagementClient`:
  `SampleApp azsdk-js-azureservicebus/7.0.0-preview.5 core-http/1.1.5 Node/v12.16.0 OS/(x64-Windows_NT-10.0.18363)`
- Added `deadLetterErrorDescription` and `deadLetterReason` properties on the received messages. Previously, they were under the `properties` in the message.

  OLD: `message.properties["DeadLetterReason"]` and `message.properties["DeadLetterErrorDescription"]`  
  NEW: `message.deadLetterReason` and `message.deadLetterErrorDescription`

- Added tracing support to the methods under `ServiceBusManagementClient`.

##### Bug Fixes

> - Fixes [bug 9926](https://github.com/Azure/azure-sdk-for-js/issues/9926)

Reviewer 4: Let's skip this one. The linked bug was fixed in Track 1 and what we did here is to port it over to Track 2. It does not need to be called out as it falls under the category of "general improvements" and there have been quite a few of these that we did

#### GA

- Identity
- Azure Key Vault

#### Updates

- Core libraries
- Azure Event Hubs
- Azure Form Recognizer

#### Preview

- Azure Service Bus

## Installation Instructions

To install the packages, copy and paste the below into a terminal.

```bash
$> npm install @azure/identity
$> npm install @azure/eventhubs-checkpointstore-blob
$> npm install @azure/ai-form-recognizer
$> npm install @azure/keyvault-keys
$> npm install @azure/keyvault-secrets
$> npm install @azure/keyvault-certificates
$> npm install @azure/service-bus@next
```

## Feedback

If you have a bug or feature request for one of the libraries, please post an issue at the [azure-sdk-for-js repository](https://github.com/azure/azure-sdk-for-js/issues).

## Changelog

Detailed changelogs are linked from the [Quick Links](#quick-links) below. Here are some of the highlights:

### Identity

#### Identity [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

> 

Reviewer 4: ```suggestion

We are happy to announce that the features we had introduced as a preview over the past few months are now stable and out of preview!

```

##### New Features
- With 1.1.0, new developer credentials are now available: `VisualStudioCodeCredential` and `AzureCliCredential`.
  - `VisualStudioCodeCredential` allows developers to authenticate using the credentials available after logging in through the Azure Account extension in Visual Studio Code.
  - `AzureCliCredential` allows developers to log into Azure using the login credentials after an "az login" call.
  - Both `VisualStudioCodeCredential` and `AzureCliCredential` may be used directly or indirectly as part of `DefaultAzureCredential`.
- Added the ability to configure the Managed Identity with a user-assigned client ID via a new option available in the `DefaultAzureCredential` constructor options: `managedIdentityClientId`.
- A list of known authorities is now available via a new top-level constant: `AzureAuthorityHosts`.
- Introduced the `CredentialUnavailable` error, which allows developers to differentiate between a credential not being available and an error happening during authentication.

### Azure Form Recognizer

> 

Reviewer 4: ```suggestion

We hav yet another preview with some API changes for a generally improved experience and which targets the service version 2.0
```

##### Breaking Changes

- Renamed the `includeSubFolders` property of the `TrainSourceFilter` type to `includeSubfolders`.
- Renamed the `documentName` property of the `TrainingDocumentInfo` type to just `name`.
- Removed the `containingLine` property of the `FormWord` type.
- Renamed `CustomFormField` to `CustomFormModelField` for similarity to other language SDKs.
- Removed the redundant `expirationDateTimeTicks` property from the `CopyAuthorization` type, as the `expiresOn` property exists.
- Moved the optional `contentType` parameter of the `FormRecognizerClient` recognition methods (`recognizeContent`, `recognizeCustomForms`, `recognizeReceipts`, and their URL-based variants) to the associated options bag for these methods.
- Removed exports of several internal types, including most internal poller operation states and some unused types. All client poller implementations now return a smaller subset of fields.

##### New Features

- Switched from using the service endpoint version `2.0-preview` to the now generally-available version `2.0`.

### Azure Key Vault Keys, Secrets and Certificates

> 

Reviewer 4: ```suggestion

We are happy to announce that the features we had introduced as a preview a few months ago are now stable and out of preview!
```

- [Key Vault Certificates](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-certificates/CHANGELOG.md).

##### New Features on Keys, Secrets and Certificates

- Added a `serviceVersion` property to the `CertificateClient`, `SecretClient`, `KeyClient` and `CryptographyClient` optional parameters to control the version of the Key Vault service being used by the clients.
    - It defaults to the latest supported API version, which currently is `7.1`.
    - The other supported service version at the moment is `7.0`.
- Added `recoverableDays` as an optional property to `KeyProperties`, `SecretProperties` and `CertificateProperties`, which denotes the number of days in which the Key, Secret or Certificate can be recovered after deletion.
    - This is only applicable for Azure Key Vaults with the `soft-delete` setting enabled.

##### New Features on Key Vault Keys

- Added `import` to the list of possible values for `KeyOperation`.

##### Major Fixes on Key Vault Keys

- Fixed [bug 10352](https://github.com/Azure/azure-sdk-for-js/issues/10352), which caused the cryptography operations on `RSA-HSM` keys to fail.

### Azure Service Bus

> 

Reviewer 4: ```suggestion

We have another preview for you containing API changes for a generally improved user experience.
```

Detailed changelogs are linked from the [Quick Links](#quick-links) below. Here are some of the highlights:

### Identity

#### [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

##### New Features

> - Extended DefaultAzureCredential with an experimental credential that uses the login credential from Azure CLI.
> - Extended DefaultAzureCredential with an experimental credential that uses the login credential from VSCode's Azure Account extension.

Reviewer 3: ```suggestion
- Extended `DefaultAzureCredential` with experimental credentials that use login credentials from Azure CLI or Visual Studio Code's Azure Account extension.
```

- Add the ability to read AZURE_AUTHORITY_HOST from the environment.
- Made all the developer credentials public as well as the list of credentials used by DefaultAzureCredential.

> - Make the keytar dependency optional, allowing for building and running on platforms not supported by keytar.

Reviewer 3: keytar change should be under Fixes?

Author: I'll have an answer soon.

Author: I will be able to update the identity section next week.

##### Breaking Changes

[TODO]

##### New Features

[TODO]

##### Major Fixes

[TODO]

### Azure Form Recognizer

#### Form Recognizer's [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/formrecognizer/ai-form-recognizer/CHANGELOG.md)

##### Breaking Changes

[TODO]

##### New Features

[TODO]

##### Major Fixes

[TODO]

### Azure Key Vault

#### Key Vault Keys' [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-keys/CHANGELOG.md)

##### Breaking Changes

[TODO]

##### New Features

[TODO]

##### Major Fixes

[TODO]

#### Key Vault Secrets' [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-secrets/CHANGELOG.md)

##### Breaking Changes

[TODO]

##### New Features

[TODO]

##### Major Fixes

[TODO]

#### Key Vault Certificates' [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-certificates/CHANGELOG.md)

##### Breaking Changes

[TODO]

##### New Features

[TODO]

##### Major Fixes

[TODO]

### Azure Service Bus

#### Service Bus' [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/service-bus/CHANGELOG.md)

> ##### Breaking Changes

Reviewer 2: ```suggestion
##### Breaking Changes from Last Preview
```

Author: Would this apply only here or in every other `Breaking Changes`?

Reviewer 2: That's how it's used to be in the previous release notes, I guess to clear confusions.
Should be applicable for all preview packages.

Reviewer 3: For preview it's nice to have "from Last Preview" to make it sound less serious

> [TODO]

Reviewer 2: ```suggestion

- `receiveMode` parameter in the `createReceiver()`, `createSessionReceiver()` and `createDeadletterReceiver()` methods has been moved into the options bag with the default value `"peekLock"` mode.

  Example:

  - OLD: `createReceiver(<queue-name>, "peekLock")` and `createReceiver(<queue-name>, "receiveAndDelete")`
  - NEW: `createReceiver(<queue-name>)` and `createReceiver(<queue-name>, {receiveMode: "receiveAndDelete"})`

- Added Async iterable iterators with pagination support for all the listing methods like `getQueues()`, `getTopics()`, `getQueuesRuntimeInfo()`, etc. and renamed them to use the `list` verb.
  - Please refer to the examples in the `samples` folder - [listingEntities](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/servicebus/service-bus/samples/typescript/src/advanced/listingEntities.ts)
- `receiveMessages()`'s optional `maxWaitTimeInMs` parameter now controls how long to wait for the _first_
  message, rather than how long to wait for an entire set of messages. This change allows for a faster return of messages to your application.
- `userProperties` attribute under the `ServiceBusMessage`(and `ReceivedMessage`, `ReceivedMessageWithLock`) has been renamed to `properties`. Same change has been made to the `userProperties` attribute in the correlation-rule filter.
- Fixes [bug 9926](https://github.com/Azure/azure-sdk-for-js/issues/9926)
  where attempting to create AMQP links when the AMQP connection was in the
  process of closing resulted in a `TypeError` in an uncaught exception.

- The terms `RuntimeInfo` and `Description` are replaced with `RuntimeProperties` and `Properties` to better align with guidelines around the kind of suffixes we use for naming methods and interfaces.
```

##### New Features

> [TODO]

Reviewer 2: ```suggestion
- User agent details can now be added to the outgoing requests by passing the user-agent prefixes to the `ServiceBusClient` and the `ServiceBusManagementClient` through options.
  Example user-agent string if the prefix `SampleApp` is provided to `ServiceBusManagementClient`:
  `SampleApp azsdk-js-azureservicebus/7.0.0-preview.5 core-http/1.1.5 Node/v12.16.0 OS/(x64-Windows_NT-10.0.18363)`
- Adding `deadLetterErrorDescription` and `deadLetterReason` properties on the received messages. Previously, they were under the `properties` in the message.

  OLD: `message.properties["DeadLetterReason"]` and `message.properties["DeadLetterErrorDescription"]`
  NEW: `message.deadLetterReason` and `message.deadLetterErrorDescription`

- Adds tracing support to the methods under `ServiceBusManagementClient`.
```

> ##### Major Fixes

Reviewer 2: ```suggestion
##### Bug Fixes

- Fixes [bug 9926](https://github.com/Azure/azure-sdk-for-js/issues/9926)
  where attempting to create AMQP links when the AMQP connection was in the
  process of closing resulted in a `TypeError` in an uncaught exception.

```

- Added the ability to configure the Managed Identity with a user-assigned client ID via a new option available in the `DefaultAzureCredential` constructor options: `managedIdentityClientId`.
- Made a list of known authorities is now available via a new top-level constant: `AzureAuthorityHosts`.
- Introduced the `CredentialUnavailable` error, which allows developers to differentiate between a credential not being available and an error happening during authentication.### Azure Event Hubs

### Event Hubs

#### [Azure Event Hubs Checkpoint Store's Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/eventhubs-checkpointstore-blob/CHANGELOG.md)

##### Major Fixes

- Fixes issue [#10132](https://github.com/Azure/azure-sdk-for-js/issues/10132)
  where using an Azure Storage Account with soft-deletes or blob versioning enabled
  would cause `listCheckpoints` to suffer performance penalties proportional to the
  number of times `updateCheckpoint` was called.

### Azure Form Recognizer

#### Form Recognizer's [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/formrecognizer/ai-form-recognizer/CHANGELOG.md)

##### Breaking Changes

> [TODO]

Reviewer 5: ```suggestion
- Renamed the `includeSubFolders` property of the `TrainSourceFilter` type to `includeSubfolders`.
- Renamed the `documentName` property of the `TrainingDocumentInfo` type to just `name`.
- Removed the `containingLine` property of the `FormWord` type.
- Renamed `CustomFormField` to `CustomFormModelField` for similarity to other language SDKs.
- Removed the redundant `expirationDateTimeTicks` property from the `CopyAuthorization` type, as the `expiresOn` property exists.
- Moved the optional `contentType` parameter of the `FormRecognizerClient` recognition methods (`recognizeContent`, `recognizeCustomForms`, `recognizeReceipts`, and their URL-based variants) to the associated options bag for these methods.
- Removed exports of several internal types, including most internal poller operation states and some unused types. All client poller implementations now return a smaller subset of fields.

```

Author: Thank you!

##### New Features

> [TODO]

Reviewer 5: ```suggestion
- Changed the package version to 3.0.0-preview.1 to reduce confusion with older versions of the Azure Form Recognizer SDKs.
- Switched to using the generally-available 2.0 service endpoint rather than 2.0-preview.
- Changed the type of the `options` bag parameter of `beginRecognizeReceipts` and `beginRecognizeReceiptsFromUrl` to `BeginRecognizeReceiptsOptions`.
- Added a `pageNumber` property to the `FormTable` and `FormTableCell` types indicating the number of the page where the table/cell appeared within the input document.
- Made the `rowSpan`, `columnSpan`, `isHeader`, and `isFooter` properties of the `FormTableCell` type non-optional to reflect that they have default values.
```

Author: Thank you!

> ##### Major Fixes
> 
> [TODO]

Reviewer 5: ```suggestion
```

Author: Thank you!

### Identity

#### [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

##### New Features
- With 1.1.0, new developer credentials are now available: `VisualStudioCodeCredential` and `AzureCliCredential`.
  - `VisualStudioCodeCredential` allows developers to log into Azure using the credentials available after logging in through the Azure Account extension in Visual Studio Code.
  - `AzureCliCredential` allows developers to log into Azure using the login credentials after an "az login" call.
- Both `VisualStudioCodeCredential` and `AzureCliCredential` may be used directly or indirectly as part of `DefaultAzureCredential`.
- Added the ability to configure the Managed Identity with a user-assigned client ID via a new option available in the `DefaultAzureCredential` constructor options: `managedIdentityClientId`.
- Made a list of known authorities is now available via a new top-level constant: `AzureAuthorityHosts`.
- Introduced the `CredentialUnavailable` error, which allows developers to differentiate between a credential not being available and an error happening during authentication.### Azure Event Hubs

> #### [Azure Event Hubs Checkpoint Store's Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/eventhubs-checkpointstore-blob/CHANGELOG.md)

Reviewer 3: ```suggestion
### Event Hubs

#### [Azure Event Hubs Checkpoint Store's Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/eventhubs-checkpointstore-blob/CHANGELOG.md)
```

##### Major Fixes

- Fixes issue [#10132](https://github.com/Azure/azure-sdk-for-js/issues/10132)
  where using an Azure Storage Account with soft-deletes or blob versioning enabled
  would cause `listCheckpoints` to suffer performance penalties proportional to the
  number of times `updateCheckpoint` was called.

### Azure Form Recognizer

#### Form Recognizer's [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/formrecognizer/ai-form-recognizer/CHANGELOG.md)

##### Breaking Changes

[TODO]

##### New Features

[TODO]

##### Major Fixes

[TODO]

### Azure Key Vault

#### Key Vault Keys' [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/keyvault/keyvault-keys/CHANGELOG.md)

##### New Features

> - Added the optional `serviceVersion` property to the `KeyClient` and `CryptographyClient` optional parameters to control the version of the Key Vault service being used by the clients.

Reviewer 3: A large portion are common among key vault libraries.  Can we combine them under

Keys, Certificates, Secrets

then list things unique to specific lib?

Author: can do! thank you

Detailed changelogs are linked from the [Quick Links](#quick-links) below. Here are some of the highlights:

### Identity

#### [Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/identity/identity/CHANGELOG.md)

##### New Features
- Extended `DefaultAzureCredential` with experimental credentials that use login credentials from Azure CLI or Visual Studio Code's Azure Account extension.
- Add the ability to read AZURE_AUTHORITY_HOST from the environment.
- Made all the developer credentials public as well as the list of credentials used by DefaultAzureCredential.
- Make the keytar dependency optional, allowing for building and running on platforms not supported by keytar.

### Azure Event Hubs

#### [Azure Event Hubs Checkpoint Store's Changelog](https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/eventhub/eventhubs-checkpointstore-blob/CHANGELOG.md)

> ##### Breaking Changes

Reviewer: ```suggestion
```

(go ahead and delete everything but #### Major Fixes)

Author: Did you mean only in this library? I just did that. Let me know if you meant in other places too.

[TODO]

##### New Features

[TODO]

##### Major Fixes

> [TODO]

Reviewer: ```suggestion
- Fixes issue [#10132](https://github.com/Azure/azure-sdk-for-js/issues/10132)
  where using an Azure Storage Account with soft-deletes or blob versioning enabled
  would cause `listCheckpoints` to suffer performance penalties proportional to the
  number of times `updateCheckpoint` was called.
```
